[Nihal Ramesh] iP

nihalramesh opened the pull request:
**Meet LeBron**

- The Goat 🐐 
- Best ChatBot
- Easy to Learn

To use it,

1. _Download It from [here](https://github.com/nihalramesh12629/ip)_
2. Done!

Heres the Java `main` Code, if you would like to see 
```java
public class Main {
    public static void main(String[] args) {
        new LeBron("./data/lebron.txt").run();
    }
}
```

Features 👽👽
- [ ] Manages your tasks
- [ ] Manages your deadlines
- [ ] Manages your events

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

Commit:
Level-0 Completed

Commit:
Level-1 Completed

Commit:
Level-2 Completed

Commit:
Level-3 Completed

Commit:
Level-4 Completed

Commit:
A-TextUiTesting Completed

Commit:
Level-5 Completed

Commit:
Level-6 Completed

Commit:
Add saving capability

Commit:
Level-7 Completed

Commit:
Merge branch-Level-7 into master

Commit:
Add LocalDate capability

Commit:
Merge branch-Level-8 into master

Commit:
Add TaskList class

Commit:
Add Ui and Storage class

Commit:
Implement further OOP into classes

Commit:
Implement packages into LeBron

Commit:
Merge remote-tracking branch 'origin/add-gradle-support'

Commit:
Add Gradle to Chatbot

Commit:
Add JUnit Tests

Commit:
Add Jar File

Commit:
Add Java Docs

Commit:
Improve Coding Standard

Commit:
Add finding capabilities

Commit:
Merge improved coding standards

Commit:
Merge finding capabilities

Commit:
Add GUI into Chatbot

Commit:
Merge branch 'branch-Level-10'

juliantayyc commented on data/lebron.txt:
Should this be in .gitignore?

juliantayyc commented on src/main/java/lebron/Parser.java:
```diff
@@ -0,0 +1,54 @@
+            ToDos todo = new ToDos(words[1].trim());
+            return new AddCommand(todo);
+        case "deadline":
+            String[] splStr = words[1].split("/by ", 2);
```
Perhaps `splStr` should be changed to a more descriptive variable name?

muhdbhz replied:
I second this 👍

juliantayyc commented on src/main/java/lebron/TaskList.java:
```diff
@@ -0,0 +1,57 @@
+        if (index <= this.tasks.size()) {
+            this.tasks.remove(index - 1);
+        } else {
+            throw new LeBronException("Thats out of bounds bro");
```
Very engaging exception message!

juliantayyc commented on src/main/java/lebron/Event.java:
```diff
@@ -0,0 +1,73 @@
+     * the task type, status icon, description, and formatted start and end
+     * dates.
+     *
+     * @return A string representing the event task.
```
Maybe can consider including an example of how the string representation might look like?

nihalramesh replied:
Aight Cheers Bro!

juliantayyc commented on src/main/java/lebron/MarkCommand.java:
```diff
@@ -0,0 +1,20 @@
+    public String execute(TaskList taskList, Ui ui, Storage storage) throws LeBronException {
+        Task task = taskList.getTask(index);
+        taskList.markTask(index);
+        String msg = ui.showTaskMarked(task);
```
You could consider using the full name for `msg` to avoid "texting style" variable names, perchance?

juliantayyc reviewed (commented):
LGTM. Just a few nits to fix. Well done!

muhdbhz commented on src/main/java/lebron/Main.java:
```diff
@@ -0,0 +1,33 @@
+ */
+public class Main extends Application {
+
+    private static final String DEFAULT_FILE_PATH = "./data/lebron.txt";
```
Nice that you capitalised the name of your constant!

muhdbhz commented on src/main/java/lebron/LeBron.java:
```diff
@@ -0,0 +1,63 @@
+ * initialization of the application, including loading tasks, handling user
+ * commands, and controlling the application's flow.
+ */
+public class LeBron {
```
Lovely class name!

muhdbhz reviewed (commented):
LGTM! Naming convention seems to be good.

Commit:
Improve code quality by refactoring Event

Refactored the Event class to enhance code readability,
maintainability and abstraction. The following changes
were made:

- Extracted redundant code blocks into parseEventDate method
- Improved method naming for better clarity and self-documentation

These improvements aim to make the codebase cleaner, easier to understand,
and more maintainable, reducing the likelihood of bugs and making
future enhancements easier.

Commit:
Add assertions to chatbot

Commit:
Add Assertions to Event

Commit:
Add Assertions to Event

nihalramesh committed:
Merge pull request #2 from nihalramesh12629/branch-A-Assertions

Add assertions to Chatbot

Commit:
Merge branch 'branch-A-CodeQuality' into master to improve code quality

This merge incorporates changes from `branch-A-CodeQuality` into the
`master` branch to improve the overall code quality of the project.
The following enhancements and refactorings have been made:

- Refactored `Event` to adhere to SOLID principles.
- Extracted duplicated code blocks into reusable helper methods to
  reduce redundancy.
- Improved method naming for better clarity and readability.
- Simplified complex conditional statements and control flows for
  better maintainability.
- Ensured proper encapsulation by making fields and methods private
  where appropriate.

These improvements aim to make the codebase more readable, maintainable,
and scalable, reducing the likelihood of bugs and making future
enhancements easier to implement.

nihalramesh committed:
Merge pull request #1 from nihalramesh12629/branch-A-CodeQuality

Improve code quality by refactoring Event

Commit:
Add functionality to snooze, postpone, and reschedule tasks

This commit introduces a new feature that allows users to easily
snooze, postpone, or reschedule tasks in the application. The following
changes have been made to implement this feature:

- Added `RescheduleCommand` class to handle rescheduling logic for tasks.
- Updated `Task`, `Deadlines`, and `Event` classes to support changing
  due dates and start/end dates with new `reschedule` methods.
- Modified the `Parser` class to parse user input for reschedule commands
  and create appropriate `RescheduleCommand` instances.
- Enhanced the `Ui` class to provide user feedback when tasks are
  successfully rescheduled or when errors occur (e.g., invalid date format
  or invalid task index).
- Ensured that the `Storage` class saves the updated task list after
  rescheduling tasks.

With these changes, users can now use commands like `reschedule` to
modify the dates of tasks and events in a user-friendly manner. This
improves the flexibility and usability of the application.

Commit:
Add functionality to snooze, postpone, and reschedule tasks

This commit introduces a new feature that allows users to easily
snooze, postpone, or reschedule tasks in the application. The following
changes have been made to implement this feature:

- Added `RescheduleCommand` class to handle rescheduling logic for tasks.
- Updated `Task`, `Deadlines`, and `Event` classes to support changing
  due dates and start/end dates with new `reschedule` methods.
- Modified the `Parser` class to parse user input for reschedule commands
  and create appropriate `RescheduleCommand` instances.
- Enhanced the `Ui` class to provide user feedback when tasks are
  successfully rescheduled or when errors occur (e.g., invalid date format
  or invalid task index).
- Ensured that the `Storage` class saves the updated task list after
  rescheduling tasks.

With these changes, users can now use commands like `reschedule` to
modify the dates of tasks and events in a user-friendly manner. This
improves the flexibility and usability of the application.

Commit:
Merge branch 'branch-BCD-Extension'

Commit:
Add README for LeBron James ChatBot User Guide

Added a comprehensive README file for the LeBron James ChatBot project, detailing the usage and features of the chatbot. The README includes:

- Introduction and product overview that incorporates LeBron James-themed jargon.
- Instructions on adding deadlines with examples and expected outcomes.
- Instructions on adding events with usage examples and outcomes.
- Detailed descriptions of the core task management features such as adding, deleting, marking, and unmarking tasks.
- Feature descriptions for searching tasks (find) and listing all tasks (list), with usage examples.
- Explanation of the 'stats' feature for monitoring productivity.
- Encouraging language that aligns with LeBron James' basketball persona to make the user guide engaging and thematic.

This README will serve as a user guide for new users and contributors to understand the chatbot's functionality and how to use its features effectively.

Commit:
Add README for LeBron James ChatBot User Guide

Added a comprehensive README file for the LeBron James ChatBot project, detailing the usage and features of the chatbot. The README includes:

- Introduction and product overview that incorporates LeBron James-themed jargon.
- Instructions on adding deadlines with examples and expected outcomes.
- Instructions on adding events with usage examples and outcomes.
- Detailed descriptions of the core task management features such as adding, deleting, marking, and unmarking tasks.
- Feature descriptions for searching tasks (find) and listing all tasks (list), with usage examples.
- Explanation of the 'stats' feature for monitoring productivity.
- Encouraging language that aligns with LeBron James' basketball persona to make the user guide engaging and thematic.

This README will serve as a user guide for new users and contributors to understand the chatbot's functionality and how to use its features effectively.

nihalramesh committed:
Update README.md

Commit:
Improve code quality by renaming boolean variables for clarity

- Renamed boolean variables to follow best practices for boolean naming conventions, improving readability and maintainability.
  - Changed `done` to `isDone`

These changes enhance the code's readability by ensuring that boolean variable names clearly convey their state (e.g., using "is", "has", "can"). This refactoring is part of an ongoing effort to improve overall code quality and consistency.

Commit:
Improve code quality by renaming boolean variables for clarity

- Renamed boolean variables to follow best practices for boolean naming conventions, improving readability and maintainability.
  - Changed `done` to `isDone`

These changes enhance the code's readability by ensuring that boolean variable names clearly convey their state (e.g., using "is", "has", "can"). This refactoring is part of an ongoing effort to improve overall code quality and consistency.

Commit:
Revise and improve README for LeBron James ChatBot

- Refactored the README file to enhance clarity and readability.
- Fixed formatting issues and added missing details for certain commands.
- Ensured that all usage examples have corresponding expected outcomes for better user understanding.

These changes improve the documentation quality and provide clearer guidance to users, aligning the README with best practices for user guides and enhancing the overall onboarding experience.

Commit:
Add error handling for missing arguments in Parser class

- Implemented a try-catch block in the parse method of the Parser class
- Added handling for ArrayIndexOutOfBoundsException when commands that require arguments (e.g., delete, mark, unmark, find, etc.) are missing additional input (words[1])
- Throws a LeBronException with a user-friendly message ("The command is missing arguments, bro!") when arguments are not provided
- Ensured the application handles missing command arguments gracefully, improving user experience and error feedback
